### PR TITLE
Save progress and login

### LINKED ISSUE #31: Add account access to the `NavBar` component
image

image

The nav bar should show a blank avatar icon when logged out, and the avatar image the user has picked when logged in.

Clicking this icon should show the account overlay to log in or out.

### LINKED ISSUE #32: Make avatars selectable in signup modal
image

The user should be able to pick one of those avatars to identify them in the future.

This modal component is currently called `SaveProgressButton`, which is not quite right. I'd rename to to `SignupModal` or something like it.

[Design in Figma](https://www.figma.com/file/LqjK3Tpvd9KJ4buFArCJBQ/Saving-Satoshi?node-id=393%3A10917&t=Ir9L2sg7RUjtqKyc-1).

### LINKED ISSUE #33: Develop the login modal
This allows users to log back in and load their progress using their personal code ([designs](https://www.figma.com/file/LqjK3Tpvd9KJ4buFArCJBQ/Saving-Satoshi?node-id=393%3A10941&t=Ir9L2sg7RUjtqKyc-1)).

image

### LINKED ISSUE #36: Check that personal code (private key) is within secp256k1 range
Users are able to save their progress with a personal code (aka a private key) that is generated in [`components/chapters/SaveProgressButton.tsx`](https://github.com/saving-satoshi/saving-satoshi/blob/master/components/chapters/SaveProgressButton.tsx#L41)

Screenshot from 2022-12-16 14-04-18

This key is a random 64 digit hex number (aka 256 bits).

As a user, I would like to use this code to generate addresses of my own in chapter 3. In order for this to work, the secret key must be smaller than 1.1578 * 10^77 (a number that is slightly less than 2^256). 

I'm not sure what the chances are of a random 256 bit number being larger than the acceptable range. That will probably determine the priority of this ticket. However, if we don't add a check, it will result in some kind of instability in the chapter 3 challenges, because those challenges have the user generate a public key (aka a point on the secp256k1 elliptic curve) from this secret key. 

Possibly a good first issue?

### PR DESCRIPTION
**Pull request by aureleoules**
Fixes #32.
Fixes #36.
Fixes #33.
Fixes #31 
Addresses #1.

- Moved `SaveProgressButton.tsx` to `SignUpModal.tsx`
- Avatar is now selectable
- Callback function to retrieve the selected avatar and code
- Generate secp256k1 private key instead of generic random hex number
- Store progress, avatar and public key in localStorage
- Login modal 
- Added avatar in Navbar

## Pull Request checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build was run locally and any changes were pushed
- [ ] Lint has passed locally and any fixes were made for failures

### CODE REVIEW

File: components/chapters/LoginModal.tsx
Comment (GBKS) [inline components/chapters/LoginModal.tsx:10]: Let's use Tailwind instead of these custom styles.

File: components/chapters/LoginModal.tsx
Comment (GBKS) [inline components/chapters/LoginModal.tsx:96]: This should probably be the current users avatar and not always the same one.

File: components/chapters/LoginModal.tsx
Comment (GBKS) [inline components/chapters/LoginModal.tsx:103]: How can we determine the users progress? Since we also need to show this on the chapters overview page (and probably elsewhere), is it best to create a utility class for this?

File: components/chapters/SignUpModal.tsx
Comment (GBKS) [inline components/chapters/SignUpModal.tsx:22]: Also change to Tailwind.

File: components/chapters/SignUpModal.tsx
Comment (GBKS) [inline components/chapters/SignUpModal.tsx:57]: Eventually we want proper aria attributes for buttons to make the UI accessible. For now, a title tag should be enough to explain what this button does..

File: components/ui/NavBar.tsx
Comment (GBKS) [inline components/ui/NavBar.tsx:33]: The `<ul>` tag can be a `<nav>`, and then we don't need the `<li>` tags. A cleaner structure.

File: components/ui/NavBar.tsx
Comment (GBKS) [inline components/ui/NavBar.tsx:47]: Should be a `<button>` tag since it's clickable.

File: components/ui/NavBar.tsx
Comment (GBKS) [inline components/ui/NavBar.tsx:67]: Should also be a `<button>` tag. Also, we need to figure out what happens when I am logged in and click the avatar - maybe another modal to show my key and the option to log out?

**Review by GBKS — CHANGES_REQUESTED** [review]
Just a few points I came across when reviewing. I need to design another modal state for when you're logged in already, and the error state for when your key doesn't match.

File: components/chapters/LoginModal.tsx
Comment (aureleoules) [reply components/chapters/LoginModal.tsx]: I've updated it to show the user's progress.

image
But I don't think this is ideal, we should probably have a utility class for this in deed.

File: components/ui/NavBar.tsx
Comment (aureleoules) [reply components/ui/NavBar.tsx]: > another modal to show my key and the option to log out?

Yes I agree.

File: components/chapters/SignUpModal.tsx
Comment (satsie) [inline components/chapters/SignUpModal.tsx]: This isn't really actionable feedback, but I want to make note (really a note to myself) of storing the public key. In chapter 3 I have an open question about manually deriving the public key from the private key. 

Recall that the public key is actually a point on the elliptic curve. Public key = private key * Generator point

I believe for ECDSA signatures, the public key is 1 byte (indicating if the y coordinate is even or odd) plus 32 bytes for the x coordinate in full.

However for Schnorr, I am reading that the public key is just the x coordinate (32 bytes). Someone please correct me if my understanding is wrong!

This all may lead to a slight discrepancy in what happens in lesson 3 and what we have stored for the public key. Anyways this is not really related to PR review, but something I need to keep in mind!

File: components/chapters/SignUpModal.tsx
Comment (satsie) [inline components/chapters/SignUpModal.tsx:17]: Thanks for adding this check! I would assume that the library you are using has this covered since it's specific to secp256k1, but I think it's still a good idea to have this safety net, just to be double sure.

**Comment by satsie** [comment]
While I can't speak to any of the front end stuff, I am able to ACK the public/private key login stuff for 83a13ed

File: components/chapters/SignUpModal.tsx
Comment (aureleoules) [reply components/chapters/SignUpModal.tsx]: Public keys for ECDSA and Schnorr are both an X and a Y coordinate (32 bytes each).
But yes you are correct, in Bitcoin ECDSA public keys are compressed into 33 bytes: the X coordinate and the parity byte (whether Y is odd or not)
For Schnorr the compressed key is 32 bytes, only the X coordinate is kept and the tie-breaker is implicit.

For lesson 3, since we are deriving ECDSA keys, the user should decide what parity byte to use (either `0x02` if Y is even, otherwise `0x03`. Or we let the user generate uncompressed keys.

File: components/chapters/SignUpModal.tsx
Comment (aureleoules) [reply components/chapters/SignUpModal.tsx:17]: I originally thought this too but after looking at the code (https://github.com/lionello/secp256k1-js/blob/master/src/secp256k1.js), the curve order doesn't seem to be checked so I added it.

File: components/chapters/SignUpModal.tsx
Comment (satsie) [reply components/chapters/SignUpModal.tsx]: oooo thank you for the clarification! Would there be any reason to let the user generate a Schnorr public key? 

I am leaning towards ending the lesson at the uncompressed keys, just to avoid this altogether :sweat_smile:

Comment (aureleoules) [reply components/chapters/SignUpModal.tsx]: > Would there be any reason to let the user generate a Schnorr public key?

I would love to support Schnorr but I believe it is much more complex to derive a P2TR address from a Schnorr public key than a legacy P2PKH from an ECDSA public key. So that may not be very fun for the user.

It shouldn't be too much trouble to ask the user which parity byte to use. Maybe allow the user to run `Y % 2` in the terminal to check if Y is odd or not, or let the user decide randomly until the challenge accepts the solution.

**Review by satsie — APPROVED** [review]

**Comment by satsie** [comment]
@GBKS gentle reminder to approve this if you feel the changes you requested have been resolved. While we can bypass the branch protections on the merge, I'd like to avoid if possible :)

File: components/chapters/SignUpModal.tsx
Comment (adamjonas) [reply components/chapters/SignUpModal.tsx]: I think we stick with uncompressed to start and work our way down in complexity.

**Comment by aureleoules** [comment]
I added a few more states, based on the figma file. https://www.figma.com/file/LqjK3Tpvd9KJ4buFArCJBQ/Saving-Satoshi?node-id=1112%3A14225&t=8GV6lHISj7LrNNnh-0

Logged in:
image
image

Load button disabled when no input:
image

Incorrect key:
image

File: components/chapters/LoginModal.tsx
Comment (GBKS) [inline components/chapters/LoginModal.tsx:117]: This can be a BoxButton.

File: components/chapters/LoginModal.tsx
Comment (GBKS) [inline components/chapters/LoginModal.tsx:80]: In another comment, you had mentioned an `Icon` class. Could we use this approach here?

File: components/chapters/LoginModal.tsx
Comment (GBKS) [inline components/chapters/LoginModal.tsx:107]: Why do you need these wrapper div tags around every UI element?

File: components/chapters/LoginModal.tsx
Comment (GBKS) [inline components/chapters/LoginModal.tsx:141]: Is it worth putting creating an avatar component? We'll show it quite often in the mining lesson in a few sizes and could bake those options in.

File: components/chapters/LoginModal.tsx
Comment (GBKS) [inline components/chapters/LoginModal.tsx:156]: This can be a BoxButton. May need to add a "success" or "clicked" state though.

File: components/chapters/LoginModal.tsx
Comment (GBKS) [inline components/chapters/LoginModal.tsx:163]: This can be a BoxButton.

File: components/chapters/LoginModal.tsx
Comment (GBKS) [inline components/chapters/LoginModal.tsx:172]: This can be a BoxButton.

File: components/chapters/LoginModal.tsx
Comment (GBKS) [inline components/chapters/LoginModal.tsx:190]: This can be a BoxButton.

File: components/ui/NavBar.tsx
Comment (GBKS) [inline components/ui/NavBar.tsx:62]: It's a bit messy to have these clip-paths in inline SVGs. I need to clean up the vectors in Figma so we can export clean ones. Also, can we refer to SVG asset files and have them automatically pulled in during compile/build?

File: components/chapters/LoginModal.tsx
Comment (GBKS) [inline components/chapters/LoginModal.tsx:103]: The regular body text should be Nunito Semi-Bold.

File: components/chapters/LoginModal.tsx
Comment (GBKS) [inline components/chapters/LoginModal.tsx:109]: The font for code input should be Space Mono.

File: components/chapters/LoginModal.tsx
Comment (GBKS) [inline components/chapters/LoginModal.tsx:75]: Instead of hard-coding colors, we need to find a way to assign dynamic colors via CSS variables, based on the color theme of the lesson ([example](https://github.com/GBKS/bitcoin-designathon/blob/main/assets/css/variables.scss#L10)). Seems like Tailwind has [support](https://tailwindcss.com/docs/adding-custom-styles#resolving-ambiguities), but we need to figure out how to assign these variables based on the current lesson.

For example, each content .mdx file could have a meta attribute called theme. That theme could be applied to the whole page as a class. That class could then adjust the variables accordingly.

**Review by GBKS — CHANGES_REQUESTED** [review]
Great progress. There are still various things to tweak, which we can do in this PR, or separately.
- Some fonts are not right
- Should we introduce an `Icon` class or a way to reference SVG assets instead of copying SVG code inline?
- Some buttons can be `BoxButtons`
- Do we need an `Avatar` class?
- We need to find a way to use CSS variables based on the color theme of the current lesson

Let me know if you have questions, and want to get everything in this PR or merge the main functionality and then update things in new PRs.

File: components/chapters/LoginModal.tsx
Comment (aureleoules) [reply components/chapters/LoginModal.tsx:107]: I like to keep UI elements separated by divs but I can change it if you want.

**Comment by aureleoules** [comment]
Thanks for your feedback @GBKS! Yes I suggest we clean up the code in separate follow-ups.
I agree we should :
- Add an avatar class
- Add an Icon class
- Add more states to the BoxButton

**Comment by aureleoules** [comment]
I updated the fonts.

File: components/chapters/LoginModal.tsx
Comment (aureleoules) [reply components/chapters/LoginModal.tsx:80]: I'm not familiar with next and the tooling used in this project so I haven't been able to import svgs from files, which would be better than inlining svgs in a big Icon class I think.
I created #47.

Comment (GBKS) [reply components/chapters/LoginModal.tsx:80]: In #49 I added SVGR to allow us to keep SVG code in separate files. Nice and easy.

File: components/chapters/LoginModal.tsx
Comment (GBKS) [reply components/chapters/LoginModal.tsx:107]: Why?

File: components/chapters/LoginModal.tsx
Comment (GBKS) [reply components/chapters/LoginModal.tsx]: I introduced theme color variables in #45.

File: components/ui/NavBar.tsx
Comment (aureleoules) [reply components/ui/NavBar.tsx]: Done in ed38633f9b207f81ef94253542245d3e6c91529a

File: components/chapters/LoginModal.tsx
Comment (aureleoules) [reply components/chapters/LoginModal.tsx]: Done in 3d0cd738f917ab419773aee1fd932256d118f13c.

File: components/chapters/LoginModal.tsx
Comment (aureleoules) [reply components/chapters/LoginModal.tsx:140]: I have difficulties making it work with the hover class and can't override it easily.
I think a follow-up PR should improve this class to allow for other background colors and other hover states.

File: components/chapters/LoginModal.tsx
Comment (aureleoules) [reply components/chapters/LoginModal.tsx:98]: Sometimes it's easier to wrap form components with a div to add styles in css so it became a habit.

File: components/chapters/LoginModal.tsx
Comment (GBKS) [reply components/chapters/LoginModal.tsx:140]: 100%. I created issue #50 for this.

**Comment by GBKS** [comment]
Great updates. Let me know if the PR is ready for another review or if you're still working on things.

File: components/chapters/LoginModal.tsx
Comment (aureleoules) [reply components/chapters/LoginModal.tsx]: Done

**Comment by aureleoules** [comment]
Ready for another review round! I rebased on top of master to use the color variables.

**Comment by adamjonas** [comment]
Avatars not highlighting and modal not collapsing once I've copied


https://user-images.githubusercontent.com/755825/211050636-7697b356-dfbf-41e9-b36e-d2a48806abb8.mov

**Comment by aureleoules** [comment]
Weird, both work for me. Tested on firefox and google-chrome.
It looks like this
image

Do you have any errors in the console? 

@GBKS can you reproduce the issue?

File: components/ui/NavBar.tsx
Comment (GBKS) [inline components/ui/NavBar.tsx:10]: We already have avatar.svg in the repo, you can use that.

File: components/chapters/LoginModal.tsx
Comment (GBKS) [inline components/chapters/LoginModal.tsx:77]: Need to find a way to change the background color of the modal.

File: components/chapters/LoginModal.tsx
Comment (GBKS) [inline components/chapters/LoginModal.tsx:91]: The avatar here should be dynamic and use the one in the local storage.

File: components/chapters/LoginModal.tsx
Comment (GBKS) [inline components/chapters/LoginModal.tsx:94]: Need to hook up the link here to switch to the SignupModal.

File: components/chapters/LoginModal.tsx
Comment (GBKS) [inline components/chapters/LoginModal.tsx:168]: Chapter and lesson should be numbers, not IDs.

File: components/ui/NavBar.tsx
Comment (GBKS) [inline components/ui/NavBar.tsx:30]: We should navigate user to their latest lesson when clicking continue.

File: components/chapters/LoginModal.tsx
Comment (GBKS) [inline components/chapters/LoginModal.tsx:113]: `warning.svg` should use `currentColor` for color definitions, so the value can be overwritten with CSS. Makes the icons more flexible.

**Review by GBKS — APPROVED** [review]
Great work with this PR.

I left a few comments and have a follow-up list of visual tweaks around font sizes and spacing. We'll also need to switch out buttons with the updated BoxButton with the green style that I'm working on.

Let me know if you want to keep tweaking this PR, or feel free to merge it and we can tweak things in follow-ups.

File: components/chapters/LoginModal.tsx
Comment (aureleoules) [reply components/chapters/LoginModal.tsx]: Done

File: components/ui/NavBar.tsx
Comment (aureleoules) [reply components/ui/NavBar.tsx]: Done

**Comment by aureleoules** [comment]
I addressed some of the nits, I suggest doing the rest as follow-ups.

**Pull request merged**
